File: 0020-valid-parentheses/0020-valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val n = s.length
        if(n%2 == 1) return false;
        val st = ArrayDeque<Char>();
        for (ch in s){
            when(ch){
                '[','{','(' -> st.addLast(ch)
                '}',']',')' ->{
                    if(st.isEmpty()) return false;
                    val top = st.removeLast();
                    if(( ch ==')' && top !='(') ||
                    ( ch =='}' && top !='{') ||
                    ( ch ==']' && top !='[')
                    ) return false;
                }
            }
        }
        return st.isEmpty()
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        int n = s.size();
        if(n%2==1) return false;
        stack<int> st;
        for(auto ch : s){
            // if(ch =='(' ||ch =='[' || ch =='{' )
            //     st.push(ch);
            // else{
            //     if(st.empty()) return false;
            //     char top = st.top();
            //     st.pop();
            //     if(( ch ==')' && top !='(') ||
            //         ( ch =='}' && top !='{') ||
            //         ( ch ==']' && top !='[')
            //     ) return false;
            // }
            switch(ch){
                case '{' : case '[' : case '(' :
                    st.push(ch);
                    break;
                case ')':
                    if (st.empty() || st.top() != '(') return false;
                    st.pop();
                    break;
                case '}':
                    if (st.empty() || st.top() != '{') return false;
                    st.pop();
                    break;
                case ']':
                    if (st.empty() || st.top() != '[') return false;
                    st.pop();
                    break;   
            }    
        }
        return st.empty();
    }
};